packages/build-utils/src/fs/run-user-scripts.ts

```diff
@@ -319,55 +329,84 @@ export async function walkParentDirs({
   return null;
 }
 
+function isSet<T>(v: any): v is Set<T> {
+  return v?.constructor.name === 'Set';
+}
+
 export async function runNpmInstall(
   destPath: string,
   args: string[] = [],
   spawnOpts?: SpawnOptions,
   meta?: Meta,
   nodeVersion?: NodeVersion
-) {
+): Promise<boolean> {
   if (meta?.isDev) {
     debug('Skipping dependency installation because dev mode is enabled');
-    return;
+    return false;
   }
 
   assert(path.isAbsolute(destPath));
-  debug(`Installing to ${destPath}`);
 
-  const { cliType, lockfileVersion } = await scanParentDirs(destPath);
-  const opts: SpawnOptionsExtended = { cwd: destPath, ...spawnOpts };
-  const env = opts.env ? { ...opts.env } : { ...process.env };
-  delete env.NODE_ENV;
-  opts.env = getEnvForPackageManager({
-    cliType,
-    lockfileVersion,
-    nodeVersion,
-    env,
-  });
-  let commandArgs: string[];
+  try {
+    await runNpmInstallSema.acquire();
+    const { cliType, packageJsonPath, lockfileVersion } = await scanParentDirs(
+      destPath
+    );
+
+    // Only allow `runNpmInstall()` to run once per `package.json`
+    // when doing a default install (no additional args)
+    if (meta && packageJsonPath && args.length === 0) {
+      if (typeof meta.runNpmInstallSet === 'undefined') {
+        meta.runNpmInstallSet = new Set<string>();
+      }
+      if (isSet<string>(meta.runNpmInstallSet)) {
```

Comment:
When would `meta.runNpmInstallSet` not be a `Set<string>`? In that case, should we log a warning or something?

Reply:
I don't think it can happen since the user can't set `meta`.

This whole block could probably be simplified to:


```ts
if (!isSet<string>(meta.runNpmInstallSet)) {
  meta.runNpmInstallSet = new Set<string>();
}

if (meta.runNpmInstallSet.has(packageJsonPath)) {
  return false;
}

meta.runNpmInstallSet.add(packageJsonPath);
```